### PR TITLE
T15257

### PR DESCRIPTION
**Pull request by mattdangerw**
Node bindings

https://phabricator.endlessm.com/T15257

### CODE REVIEW

File: src/engine.cc [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  data->reject.Reset();
+  delete data;
+}
+
+void EngineGetObject(const FunctionCallbackInfo<Value> &args) {
+  Isolate *isolate = args.GetIsolate ();
+  GObject *engine = GObjectFromWrapper (args[0]);
+  String::Utf8Value id (args[1]->ToString ());
+  String::Utf8Value app_id (args[2]->ToString ());
+  Local<Function> resolve = Local<Function>::Cast(args[3]);
```
Comment (ptomato) [inline src/engine.cc]: Instead of passing in `resolve` and `reject`, I think you can create a persistent `v8::Promise::Resolver` here, use it directly as the callback data, and return its associated `v8::Promise`.

File: src/engine.cc [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  Isolate *isolate = Isolate::GetCurrent ();
+  CallbackData *data = (CallbackData *)user_data;
+
+  GError *error = NULL;
+  EkncContentObjectModel *model;
+
+  if ((model = eknc_engine_get_object_finish (EKNC_ENGINE (source), result, &error))) {
+    const unsigned argc = 1;
+    Local<Value> argv[argc] = { WrapperFromGObject (isolate, G_OBJECT (model)) };
+    Local<Function>::New(isolate, data->resolve)->Call(isolate->GetCurrentContext()->Global(), argc, argv);
```
Comment (ptomato) [inline src/engine.cc]: I may be looking at the wrong version of the v8 API docs because I see a different API for this function. Anyhow, is `isolate->GetCurrentContext()->Global()` being used as the `this` object for the function call? Does that override any bound `this` from an arrow function?

File: src/engine.cc [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+{
+  Isolate *isolate = Isolate::GetCurrent ();
+  CallbackData *data = (CallbackData *)user_data;
+
+  GError *error = NULL;
+  EkncQueryResults *results;
+
+  if ((results = eknc_engine_query_finish (EKNC_ENGINE (source), result, &error))) {
+    const unsigned argc = 1;
+    Local<Value> argv[argc] = { WrapperFromGObject (isolate, G_OBJECT (results)) };
```
Comment (ptomato) [inline src/engine.cc]: Nothing to do with this commit, but while I was reviewing it I figured it might be good to document that `WrapperFromGObject` eats a reference

File: src/value.cc [hunk 4]
```diff
@@ -77,6 +77,16 @@ Local<Value> GIArgumentToV8(Isolate *isolate, GITypeInfo *type_info, GIArgument
         }
         break;
 
+    case GI_TYPE_TAG_GSLIST:
+        {
+            Local<Array> array = Array::New (isolate);
+            GSList *list = (GSList *)arg->v_pointer;
+            for ( ; list != NULL; list = list->next) {
+                array->Set(array->Length(), WrapperFromGObject (isolate, (GObject *)list->data));
```
Comment (ptomato) [inline src/value.cc:85]: It assumes transfer full, exclusively, which may do for now as you say in the commit message. But I think it leaks the `GSList`, you have to handle `GI_TYPE_TAG_GSLIST` in `FreeGIArgument()`.

File: src/value.cc [hunk 5]
```diff
@@ -280,6 +280,26 @@ void V8ToGValue(GValue *gvalue, Local<Value> value) {
         g_value_set_enum (gvalue, value->Int32Value ());
     } else if (G_VALUE_HOLDS_OBJECT (gvalue)) {
         g_value_set_object (gvalue, GObjectFromWrapper (value));
+    } else if (G_VALUE_HOLDS_VARIANT (gvalue)) {
+        // FIXME: For now, just special case string arrays. Extending to a
+        // general gvariant parser would require the gvariant type in this
+        // function.
+        if (!value->IsArray ())
```
Comment (ptomato) [inline src/value.cc]: `g_assert (value->IsArray ())`?

File: src/value.cc [hunk 6]
```diff
@@ -305,6 +325,19 @@ Local<Value> GValueToV8(Isolate *isolate, const GValue *gvalue) {
         return Integer::New (isolate, g_value_get_enum (gvalue));
     } else if (G_VALUE_HOLDS_OBJECT (gvalue)) {
         return WrapperFromGObject (isolate, G_OBJECT (g_value_get_object (gvalue)));
+    } else if (G_VALUE_HOLDS_VARIANT (gvalue)) {
+        // FIXME: For now, just special case string arrays. For more complex
+        // gvariants we will just print out the string dump of the data.
+        GVariant *variant = g_value_get_variant (gvalue);
+        if (g_variant_is_of_type (variant, G_VARIANT_TYPE_STRING_ARRAY)) {
+            gsize length;
+            g_autofree const gchar **strings = g_variant_get_strv (variant, &length);
```
Comment (ptomato) [inline src/value.cc]: This looks weird, but it is correct :-)

File: src/value.cc [hunk 7]
```diff
@@ -305,6 +325,19 @@ Local<Value> GValueToV8(Isolate *isolate, const GValue *gvalue) {
         return Integer::New (isolate, g_value_get_enum (gvalue));
     } else if (G_VALUE_HOLDS_OBJECT (gvalue)) {
         return WrapperFromGObject (isolate, G_OBJECT (g_value_get_object (gvalue)));
+    } else if (G_VALUE_HOLDS_VARIANT (gvalue)) {
+        // FIXME: For now, just special case string arrays. For more complex
+        // gvariants we will just print out the string dump of the data.
+        GVariant *variant = g_value_get_variant (gvalue);
+        if (g_variant_is_of_type (variant, G_VARIANT_TYPE_STRING_ARRAY)) {
+            gsize length;
```
Comment (ptomato) [inline src/value.cc]: Seems like the non-`g` types are used everywhere else in the code, like `size_t` and `const char **` in this case

File: src/value.cc [hunk 8]
```diff
@@ -305,6 +325,19 @@ Local<Value> GValueToV8(Isolate *isolate, const GValue *gvalue) {
...
+            g_autofree const gchar **strings = g_variant_get_strv (variant, &length);
+            Local<Array> array = Array::New (isolate);
+            for (guint i = 0; i < length; i++)
+                array->Set (i, String::NewFromUtf8 (isolate, strings[i]));
+            return array;
+        }
+        return String::NewFromUtf8 (isolate, g_variant_print (g_value_get_variant (gvalue), FALSE));
```
Comment (ptomato) [inline src/value.cc]: Likewise, use C++ `false`

File: src/boxed.cc [hunk 9]
```diff
@@ -10,8 +10,8 @@ struct Boxed {
 
 static G_DEFINE_QUARK(gnode_js_template, gnode_js_template);
 
-static void BoxedClassDestroyed(const WeakCallbackInfo<GIBaseInfo> &cbinfo) {
-    GIBaseInfo *info = cbinfo.GetParameter ();
+static void BoxedClassDestroyed(const WeakCallbackInfo<GIBaseInfo> &data) {
+    GIBaseInfo *info = data.GetParameter ();
```
Comment (ptomato) [inline src/boxed.cc]: This ended up in the wrong commit? You changed it from `data` to `cbinfo` a couple of commits back

File: src/domain.cc [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    isolate->ThrowException (Exception::TypeError (String::NewFromUtf8 (isolate, error->message)));
+    return;
+  }
+
+  Local<Object> module_obj = Object::New (isolate);
+  if (mime)
+    module_obj->Set (String::NewFromUtf8 (isolate, "mime_type"), String::NewFromUtf8 (isolate, mime));
+  if (bytes)
+    {
+      gsize length = g_bytes_get_size (bytes);
```
Comment (ptomato) [inline src/domain.cc]: No need for this, `length` is filled in in the following line

File: src/domain.h [hunk 11]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+#include <node.h>
+
+using namespace v8;
```
Comment (ptomato) [inline src/domain.h]: Don't put `using namespace` in header files.

File: README.md [hunk 12]
```diff
@@ -1,6 +1,96 @@
...
+You can add this to any node project by running the following.
+```
+npm install --save eos-knowledge-content
+```
+
+## API
+
+For the most part, these bindings should resemble the gjs (GNOME javascript)
+bindings used in eos-knowledge-lib. Complete documentation of all objects and
+properties can be build from the
```
Comment (ptomato) [inline README.md]: build -> built

File: README.md [hunk 13]
```diff
@@ -1,6 +1,96 @@
...
+[eos-knowledge-content library](https://github.com/endlessm/eos-knowledge-lib/tree/master/ekncontent).
+
+Here's a quick start example grabbing running a query for 'foot' for an app with
```
Comment (ptomato) [inline README.md]: delete 'grabbing'

File: index.js [hunk 14]
```diff
@@ -202,6 +201,7 @@ Eknc.Engine.prototype.query = function (query) {
     });
 };
 
+// Primarily for setting up a 'ekn://' uri handler with electron
```
Comment (ptomato) [inline index.js]: Belongs in a different commit?

File: src/mainloop.cc [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+struct ThreadData {
+  uv_thread_t thread_handle;
+  uv_mutex_t mutex_handle;
+  uv_async_t async_handle;
+  GPollFD *fds;
+  gint nfds;
+};
+
+static void DispatchGLibMainloop (uv_async_t *async_handle) {
```
Comment (ptomato) [inline src/mainloop.cc]: I don't understand too well what's going on here, maybe an overview for dummies?

File: src/engine.cc [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+  Persistent<Function> reject;
+};
+
+static void EngineGetObjectCallback(GObject *source,
+                                    GAsyncResult *result,
+                                    gpointer user_data)
+{
+  Isolate *isolate = Isolate::GetCurrent ();
+  HandleScope scope(isolate);
+  CallbackData *data = (CallbackData *)user_data;
```
Comment (ptomato) [inline src/engine.cc]: In GJS I've started to use stuff like `auto data = static_cast<CallbackData *>(user_data)` so that (1) the type name is not specified twice and (2) potentially dangerous casts are made obvious by using C++ style.

File: src/bindings.cc [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+static void ObjectPropertyGetter(const FunctionCallbackInfo<Value> &args) {
+    Isolate *isolate = args.GetIsolate ();
+    GObject *gobject = GObjectFromWrapper (args[0]);
+    String::Utf8Value prop_name_v (args[1]->ToString ());
+    const char *prop_name = *prop_name_v;
+
+    GParamSpec *pspec = g_object_class_find_property (G_OBJECT_GET_CLASS (gobject), prop_name);
+    GValue value = { 0, 0 };
```
Comment (ptomato) [inline src/bindings.cc]: Use `GValue value = G_VALUE_INIT` so you don't depend on the internal structure of `GValue`

File: index.js [hunk 18]
```diff
@@ -164,30 +164,46 @@ function makeInfo(info) {
...
-    }
+    return module;
+}
 
-    if (override)
-        override.apply(module);
+const Eknc = importNS('EosKnowledgeContent');
 
-    return module;
+Eknc.start_glib_mainloop = function () {
```
Comment (ptomato) [inline index.js]: `Eknc.start_glib_mainloop = bindings.StartGLibMainloop`?

**Review by ptomato — CHANGES_REQUESTED** [review]
Nice. The main thing that I think could be improved is returning a promise directly from C++ in the async functions. Otherwise, just a few comments.

I'm having trouble testing it though:
```
$ nodejs examples/query.js 
module.js:597
  return process.dlopen(module, path._makeLong(filename));
                 ^

Error: Module version mismatch. Expected 48, got 46.
    at Error (native)
    at Object.Module._extensions..node (module.js:597:18)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/home/philip/checkout/eos-knowledge-content-node/index.js:1:80)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
```
I doubt this has anything to do with your code, but I wonder if you've seen this and know what to do about it?

File: src/engine.cc (hunk 2 above)
Comment (mattdangerw) [reply src/engine.cc]: Not sure, I think it might. But the functions here are internal to the index.js and not exposed, so I don't think we need to worry about use cases besides the one we have directly.

File: src/engine.cc (hunk 1 above)
Comment (mattdangerw) [reply src/engine.cc]: ill try that out

File: src/mainloop.cc (hunk 15 above)
Comment (mattdangerw) [reply src/mainloop.cc]: Sure. Need to update the README too with this from end user perspective.

So this is just for using purely from nodejs (not electron, where chromium uses glib's mainloop to power its event loop called MessageLoop).

The idea is to drive the usual `while (running) { poll(fds, timeout); }` glib mainloop on a separate thread from libuvs mainloop, because glibs mainloop is not easily embedable. But we still want to be able to have our glib async functions wake up on the main thread so they can calll back into v8s javascript context. So we trigger the specific part of the mainloop that will actually distpach events (`g_main_context_dispatch`) back on libuv's main thread.

Something like
```
glib thread: poll for events
glib thread: got events. wakeup libuv thread and lock
libuv thread: dispatch glib events
libuv thread: unlock glib thread
glib thread: poll for events
...
```

Credit to Jasper for the approach.

Long term, if glib moves to epoll (which newer event loops use on linux), we could just poll on a single fd from libuv and drive the glib mainloop in the same thread. Possible with the current state of things, but it would have worked out to a lot more code.

I'll put an abbreviated version of this in the comments.

File: src/bindings.cc (hunk 17 above)
Comment (mattdangerw) [reply src/bindings.cc]: oooh nice ive been looking for that...

**Comment by mattdangerw** [comment]
@ptomato as to everything not working my first suspicion would be node version. I'm on the latest `7.7` right now. Ping me if you want help getting that set up, its not much trouble at all.

Trying to keep this working with a large number of node version is probably doable, but I think best handled as part of https://phabricator.endlessm.com/T15911

File: src/engine.cc (hunk 3 above)
Comment (mattdangerw) [reply src/engine.cc]: done

File: src/engine.cc (hunk 16 above)
Comment (mattdangerw) [reply src/engine.cc]: I think I'd like to hold off on this for now. There's a lot of cleanups that could be landed, but I'd rather follow the current state of the code brought in from node-gtk for now, which is not using `static_cast`.

File: src/value.cc (hunk 4 above)
Comment (mattdangerw) [reply src/value.cc:85]: done

File: src/engine.cc [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+
+void EngineGetObject(const FunctionCallbackInfo<Value> &args) {
+  Isolate *isolate = args.GetIsolate ();
+  GObject *engine = GObjectFromWrapper (args.This ());
+  String::Utf8Value id (args[0]->ToString ());
+  String::Utf8Value app_id (args[1]->ToString ());
+
+  Local<Promise::Resolver> resolver = Promise::Resolver::New(isolate);
+  Local<Promise> promise = resolver->GetPromise();
+  CallbackData *data = new CallbackData();
```
Comment (ptomato) [inline src/engine.cc]: Can you just allocate a `Persistent<Promise::Resolver>` on the heap directly, and get rid of `CallbackData`?

File: src/engine.cc [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+                                    gpointer user_data)
+{
+  Isolate *isolate = Isolate::GetCurrent ();
+  HandleScope scope(isolate);
+  CallbackData *data = (CallbackData *)user_data;
+
+  GError *error = NULL;
+  EkncContentObjectModel *model;
+
+  Local<Promise::Resolver> resolver = Local<Promise::Resolver>::New(isolate, data->resolver);
```
Comment (ptomato) [inline src/engine.cc]: Why not use `data->resolver` directly here?

File: index.js [hunk 21]
```diff
@@ -180,4 +180,21 @@ function importNS(ns) {
     return module;
 }
 
-module.exports = importNS('EosKnowledgeContent');
+const Eknc = importNS('EosKnowledgeContent');
+
+// Setup async functions
+Eknc.Engine.prototype.get_object_for_app = function (id, app_id) {
+    return new Promise((resolve, reject) => {
+        bindings.EngineGetObject(this, id, app_id, resolve, reject);
```
Comment (ptomato) [inline index.js]: Not sure how this is working, because the `resolve` and `reject` args aren't being read any longer in `EngineGetObject()`. And IIUC you are just getting a promise returned directly from there, so I don't think you have to wrap it in another promise here.

File: src/engine.cc (hunk 19 above)
Comment (mattdangerw) [reply src/engine.cc]: Totally, but its already in and will scale better to new things we might need. So kinda like keeping the callback data is always a struct approach.

File: src/engine.cc (hunk 20 above)
Comment (mattdangerw) [reply src/engine.cc]: I'll try it. My understanding of the v8 api is pretty damn week. I only got as deep as I needed to get things working. I was trying to mirror the structure of the function code this promise code just replaced, which made a scope, then a local out of the persistent and called into that. But maybe that's not necessary at all?

File: index.js (hunk 21 above)
Comment (mattdangerw) [reply index.js]: Oh, just messed up my commit history, one sec...

Comment (mattdangerw) [reply index.js]: fixed

File: src/engine.cc (hunk 20 above)
Comment (mattdangerw) [reply src/engine.cc]: Yeah, wasn't able to get this working. And most v8 async examples I can find seem to make a local from a persistent and call that.

Comment (ptomato) [reply src/engine.cc]: Oh, I misread it initially, you are creating a new `Local<>`, not a new `Promise::Resolver`. Still don't understand why you can't call methods on the `Persistent<>`, in SpiderMonkey that works fine, but if it doesn't work, it doesn't work...

File: src/domain.cc [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+  if (!eknc_domain_read_uri (EKNC_DOMAIN (domain), *uri, &bytes, &mime, &error)) {
+    isolate->ThrowException (Exception::TypeError (String::NewFromUtf8 (isolate, error->message)));
+    return;
+  }
+
+  Local<Object> module_obj = Object::New (isolate);
+  if (mime)
+    module_obj->Set (String::NewFromUtf8 (isolate, "mime_type"), String::NewFromUtf8 (isolate, mime));
+  if (bytes)
+    {
```
Comment (ptomato) [inline src/domain.cc]: Match indentation and bracket styles with the rest of the code

File: src/domain.cc [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+
+  GObject *domain = GObjectFromWrapper (args[0]);
+  String::Utf8Value uri (args[1]->ToString ());
+
+  const char *mime = NULL;
+  GBytes *bytes = NULL;
+  GError *error = NULL;
+
+  if (!eknc_domain_read_uri (EKNC_DOMAIN (domain), *uri, &bytes, &mime, &error)) {
+    isolate->ThrowException (Exception::TypeError (String::NewFromUtf8 (isolate, error->message)));
```
Comment (ptomato) [inline src/domain.cc]: `TypeError` seems like the wrong exception to throw

File: src/boxed.cc (hunk 9 above)
Comment (ptomato) [reply src/boxed.cc]: This comment got swallowed, I think

**Review by ptomato — CHANGES_REQUESTED** [review]

File: src/engine.cc [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+  if ((model = eknc_engine_get_object_finish (EKNC_ENGINE (source), result, &error)))
+    resolver->Resolve(WrapperFromGObject (isolate, G_OBJECT (model)));
+  else
+    resolver->Reject(Exception::TypeError(String::NewFromUtf8(isolate, error->message)));
+
+  /* XXX: For some reason we have to resolve microtasks to the promise wrappers to resolve. */
```
Comment (ptomato) [inline src/engine.cc]: Probably can remove this comment, here and below. I've seen this [here](https://github.com/nodejs/node/issues/5691#issuecomment-196277918):
> It also seems that it is implied by v8 authors that you should run microtasks manually when you work with promises from C++.

Although I've been googling for a "citation needed" for that and can't find anything.

File: src/domain.cc [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+#include "gobject.h"
+
+#include <node.h>
+#include <node_buffer.h>
+#include <eos-knowledge-content.h>
+
+using namespace v8;
+
+void DomainReadURI(const FunctionCallbackInfo<Value> &args) {
+  Isolate *isolate = args.GetIsolate ();
```
Comment (ptomato) [inline src/domain.cc]: nitpick: 4 spaces to match rest of codebase

File: src/mainloop.cc [hunk 26]
```diff
@@ -0,0 +1,83 @@
+#include "mainloop.h"
+
+#include <uv.h>
+#include <glib.h>
+
+using namespace v8;
+
+struct ThreadData {
+  uv_thread_t thread_handle;
```
Comment (ptomato) [inline src/mainloop.cc]: nitpick: 4 spaces to match rest of codebase

**Review by ptomato — CHANGES_REQUESTED** [review]
Only a couple of nitpicks. I would say go ahead and commit it after fixing them, but this needs the other pull request to be merged first anyway.

File: src/engine.cc (hunk 24 above)
Comment (mattdangerw) [reply src/engine.cc]: I'll rework the comment, to say this is expected, but would like to keep something as its pretty confusing.

**Review by ptomato — APPROVED** [review]
**CR pass**
